cluster-autoscaler/core/scale_up.go

```diff
@@ -524,6 +535,80 @@ func ScaleUp(context *context.AutoscalingContext, processors *ca_processors.Auto
 	return &status.ScaleUpStatus{ScaledUp: false, PodsRemainUnschedulable: getRemainingPods(podsRemainUnschedulable, skippedNodeGroups)}, nil
 }
 
+type podsPredicatePassingCheckFunctions struct {
+	getPodsPassingPredicates    func(nodeGroupId string) ([]*apiv1.Pod, error)
+	getPodsNotPassingPredicates func(nodeGroupId string) (map[*apiv1.Pod]status.Reasons, error)
+}
+
+func getPodsPredicatePassingCheckFunctions(
+	context *context.AutoscalingContext,
+	unschedulablePods []*apiv1.Pod,
+	nodeInfos map[string]*schedulercache.NodeInfo) podsPredicatePassingCheckFunctions {
+
+	podsPassingPredicatesCache := make(map[string][]*apiv1.Pod)
+	podsNotPassingPredicatesCache := make(map[string]map[*apiv1.Pod]status.Reasons)
+	errorsCache := make(map[string]error)
+
+	computeCaches := func(nodeGroupId string) {
+		nodeInfo, found := nodeInfos[nodeGroupId]
+		if !found {
+			errorsCache[nodeGroupId] = errors.NewAutoscalerError(errors.InternalError, "NodeInfo not found for node group %v", nodeGroupId)
+		}
+
+		podsPassing := make([]*apiv1.Pod, 0)
+		podsNotPassing := make(map[*apiv1.Pod]status.Reasons)
+		schedulableOnNode := CheckPodsSchedulableOnNode(context, unschedulablePods, nodeGroupId, nodeInfo)
+		for pod, err := range schedulableOnNode {
+			if err == nil {
+				podsPassing = append(podsPassing, pod)
+			} else {
+				podsNotPassing[pod] = err
+			}
+		}
+		podsPassingPredicatesCache[nodeGroupId] = podsPassing
+		podsNotPassingPredicatesCache[nodeGroupId] = podsNotPassing
+	}
+
+	return podsPredicatePassingCheckFunctions{
+
+		getPodsPassingPredicates: func(nodeGroupId string) ([]*apiv1.Pod, error) {
+			_, passingFound := podsPassingPredicatesCache[nodeGroupId]
+			_, errorFound := errorsCache[nodeGroupId]
+
+			if !passingFound && !errorFound {
+				computeCaches(nodeGroupId)
+			}
+			pods, found := podsPassingPredicatesCache[nodeGroupId]
+			if found {
+				return pods, nil
+			}
+			err, found := errorsCache[nodeGroupId]
```

Comment:
Just in case (see the early exit comment above), I'd prefer it if we checked errorsCache first. This way, if there's an error we'll always find it, even if we accidentally set other entries for the node group. Especially given that we may e.g. want to modify cache entries for autoprovisioned node groups in the future.

Reply:
Makes sense. Updated.